docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -80,7 +79,8 @@ to provide canonical representation of a `Transaction` to sign over, clients mus
 obey the following rules:
 
 - Encode `SignDoc` (see below) via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
-  - Default and zero values must be stripped from the output (`0`, `“”`, `null`, `false`, `[]`, and `{}`).
+  - Default must be stripped from the output!
```

Comment:
Should we maybe leave `0`, `“”`, `null`, `false`, `[]`, and `{}`. in here so that it's clear?

Reply:
Well, that wouldn't necessarily be true, correct? The output can still have `[]` and `{}` for example. I suppose we can clearly stipulate that default values should be omitted.

Reply:
Well the issue is consistency. Clients should know what to do and what not to do. If  the app emits `[]` for signature verification sometimes and other times it doesn't, it's hard for clients to know how to implement signing. Knowing which cases to omit and not omit would be a lot of out of band knowledge. So the simplest is just to say omit them always, then everyone knows what to do.

Reply:
Consistency should be provided by Proto JSON. 

I'm stating that its _not_ omitted when explicitly provided, even when `[]` for example. In other words, only when the value is completely omitted, should the encoding output omit the default value.

e.g.

```go
type Foo struct {
  A int
  B []string
}
```

`protoJSON(&Foo{A: 5})` =/= `protoJSON(&Foo{A: 5, B: []string{}})`


idk...it seems like we're both stating the same thing.

Reply:
With proto 3, the idea is that empty and default are indistinguishable. So if we follow that logic, there is no correct way to distinguish empty and default.

Think about it this way, if I am a client and explicitly pass `[]` to my field in javascript, when I encode that message as protobuf binary it may or may not be omitted. When the app decodes the binary protobuf message, the field may be populated with an empty array or just null. We can't know from the protobuf binary whether the client intended to omit the field or pass `[]` because in proto 3 there is no real distinction and it may be implementation dependent. So there is no way to distinguish.

Thus I'm arguing that `protoJSON(&Foo{A: 5})` should in fact be equal to `protoJSON(&Foo{A: 5, B: []string{}})`. If it isn't clients will have a horrible time signing messages...

Reply:
@aaronc I understand, what I'm saying is Proto3 **doesn't** allow you to distinguish this. In other words, this is false:

```
With proto 3, the idea is that empty and default are indistinguishable. 
```

So either:

1. We enforce types to respect `json:"omitempty"` (which we can only do via documentation) OR
2. Use reflection on the output before passing to canonical-json-form.

Reply:
What I had been suggesting and the way I read this line in the ADR is that there are actually three passes. The first pass is proto json. The second pass is actually this stripping default values pass. Then the third pass is canonical json.

I think this second pass approach is preferable to requiring `json:"omitempty"` through documentation. It wouldn't necessarily need to be via reflection. It could be done by parsing json and reserializing without default values. Not super complex. Could be done in a separate PR from this one.

If that approach is okay, I would just revert this documentation line to make it clear for clients.

I know the second pass requires more upfront work initially, but seems better than a lot of manual coordination around `json:"omitempty"`.

What do you think?

Reply:
Yeah, OK, then, in that case, we're on the same page as I agree `omitempty` is not preferable. I believe it should be done in this PR.

However, I don't think it's super trivial. You decode Proto JSON into a `map[string][]interface{}`. You have no idea whats an empty/default value and what isn't. So unless I'm missing something, you have to check each respective value. How do you propose to "reserializing without default values"?

Reply:
Here's code that does it:

```go
func StripDefaults(val interface{}) interface{} {
	switch val := val.(type) {
	case bool:
		if val == false {
			return nil
		}
	case string:
		if val == "" {
			return nil
		}
	case float64:
		if val == 0 {
			return nil
		}
	case []interface{}:
		n := len(val)
		if n == 0 {
			return nil
		}
		res := make([]interface{}, n)
		for i, x := range val {
			res[i] = StripDefaults(x)
		}
		return res
	case map[string]interface{}:
		res := make(map[string]interface{})
		for k, v := range val {
			v2 := StripDefaults(v)
			if v2 != nil {
				res[k] = v2
			}
		}
		if len(res) == 0 {
			return nil
		}
		return res
	}
	return val
}
```

Tested it and it works so far. I can open a PR against this branch.

Reply:
See https://github.com/cosmos/cosmos-sdk/pull/5815

Reply:
Right, that's what my suggestion was (recursive check & mutation) 👍

Reply:
While this approach is similar to what I originally suggested, it's not complete or fully working (see [here](https://github.com/cosmos/cosmos-sdk/pull/5815#discussion_r393924844)). Not to mention, it's not the most ideal code and flow to maintain.

Having tried to implement a clearer alternative using reflect (which works), you still end up with `null` values that are now the final JSON output (canonical JSON doesn't omit them).

I'm starting to consider going back to the `omitempty` requirement as that's an easier approach. Thoughts?